File: 백준/Silver/6118. 숨바꼭질/숨바꼭질.py
import sys
from collections import deque

n, m = map(int, sys.stdin.readline().split())

graph = [[] for _ in range(n + 1)]

for _ in range(m):
    a, b = map(int, sys.stdin.readline().split())

    graph[a].append(b)
    graph[b].append(a)

for i in graph:
    i.sort()

dist = [0] * (n + 1)
visited = [False] * (n+1)
q = deque()

q.append(1)
visited[1] = True

while q:
    x = q.popleft()

    for i in graph[x]:
        if not visited[i]:
            dist[i] = dist[x] + 1
            visited[i] = True
            q.append(i)


print(dist.index(max(dist)),max(dist),dist.count(max(dist)))

File: 백준/Silver/6118. 숨바꼭질/숨바꼭질.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;


public class Main {

    static ArrayList<ArrayList<Integer>> arr = new ArrayList<>();
    static int[] dist;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        dist = new int[n + 1];

        for (int i = 0; i <= n; i++) {
            arr.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            arr.get(a).add(b);
            arr.get(b).add(a);
        }
        Arrays.fill(dist, -1);
        dist[1] = 0;
        for (int i = 1; i < arr.size(); i++) {
            Collections.sort(arr.get(i));
        }

        Queue<Integer> q = new LinkedList<>();

        q.offer(1);

        while (!q.isEmpty()) {
            int v = q.poll();

            for (int i = 0; i < arr.get(v).size(); i++) {
                int x = arr.get(v).get(i);
                if (dist[x] == -1) {
                    dist[x] = dist[v] + 1;
                    q.offer(x);
                }
            }
        }

        int maxDist = Arrays.stream(dist).max().getAsInt();
        int result = 0;
        int cnt = 0;
        for (int i = 0; i < dist.length; i++) {
            if (maxDist == dist[i]) {
                cnt++;
                if (result == 0) {
                    result = i;
                }
            }
        }

        System.out.printf("%d %d %d", result, maxDist, cnt);
    }
}